Invoke Local: Java Support

RafalWilinski opened the pull request:
## What did you implement:

Closes https://github.com/serverless/serverless/issues/2864

## How did you implement it:

I've created a Java subproject which handles JS <-> Java communication.

Flow:
1. User fires cmd `serverless invoke local -f hello -d '{"key":"value"}' -c '{"name":"customName"}'`
2. Serverless checks if JS <-> Java bridge is already built
3. If it's not, then it's running `mvn package` to download all the dependencies and build JAR (the "bridge").
3. Once bridge it's built, Serverless runs it with all the necessary parameters. Event & context is passed via STDIN just like in `invokeLocalPython`.
4. Bridge parses the event, loads selected function artifact dynamically using reflection and invokes function locally.

## How can we verify it:

```
docker run -it maven:3-jdk-8 /bin/bash

# Inside container
curl -sL https://deb.nodesource.com/setup_8.x | bash -
apt-get install -y nodejs

git clone -b invoke-local-maven https://github.com/RafalWilinski/serverless
cd serverless && npm install && cd ..

mkdir sls-tests
cd sls-tests

../serverless/bin/serverless create -t aws-java-maven
mvn package
../serverless/bin/serverless invoke local -f hello
```

## Remarks / to discuss:
1. In order to run that, you need maven and JDK8 installed locally.
2. In order to get correct output, developer need to implement `toString` method of `ApiGatewayResponse` class. If it's not there, output looks like this: `com.serverless.ApiGatewayResponse@42e26948`
3. I don't think that this is the right place in project structure to keep that subproject and executable. JAR might be moved to serverless/bin but what about source code?
    1. Separate repository might be good solution. We could lazy-download executable/JAR just like in run/installEventGateway. This also removes requirement of having a maven installed.

## Todos:

- [x] Write tests
- [x] Write documentation
- [x] Fix linting errors
- [x] Make sure code coverage hasn't dropped
- [x] Provide verification config / commands / resources
- [x] Enable "Allow edits from maintainers" for this PR
- [x] Update the messages below

***Is this ready for review?:*** YES
***Is it a breaking change?:*** NO

eahefnawy commented:
@RafalWilinski that's awesome! thanks a lot!

Looking through the code, don't have anything to add on the Node.js side. Unfortunately I don't have much Java experience, but I pinged our team to see if someone could take a look 😊

nikgraf commented:
@RafalWilinski you rock!

pmuens commented on .gitignore:
```diff
@@ -48,3 +48,4 @@ tracking-config.json
 
 # Misc
 jest
+lib/plugins/aws/invokeLocal/java/target
```
Should we create another `.gitignore` in the `java` directory for that?

RafalWilinski commented:
Interesting, I've followed your steps to reproduce:
```bash
docker run -it maven:3-jdk-8 /bin/bash
git clone -b invoke-local-maven https://github.com/RafalWilinski/serverless
cd serverless && npm install && cd ..
mkdir sls-tests
cd sls-tests
../serverless/bin/serverless create -t aws-java-maven
mvn package
../serverless/bin/serverless invoke local -f hello
```

And got same message, *however*, function executed properly:
```
root@f024cfe17ce6:~/sls-tests# ../serverless/bin/serverless invoke local -f hello
Building Java bridge, first invocation might take a bit longer.
mvn - POM file /root/serverless/lib/plugins/aws/invokeLocal/java specified with the -f/--file command line argument does not exist

log4j:WARN No appenders could be found for logger (com.serverless.Handler).


log4j:WARN Please initialize the log4j system properly.
log4j:WARN See http://logging.apache.org/log4j/1.2/faq.html#noconfig for more info.

com.serverless.ApiGatewayResponse@46d56d67
```

RafalWilinski commented:
@pmuens 

I've added `pom.xml` to the `mvn package -f` path explicitly and now it doesn't show that warning/error. You can try with the steps to reproduce list that I've provided here https://github.com/serverless/serverless/pull/4199#issuecomment-328582977

I've also fixed that `gitignore` suggestion.

RafalWilinski commented:
1. Sure
2. `mvn package` actually is not responsible for building bridge but for compiling your project. It's also required to run `mvn package` before `sls deploy`, that's just how Java works with Serverless framework. I can add that but it will be inconsistent with `sls deploy`, I mean:
 - `sls invoke local` - builds project for you implicitly, under the hood
 - `sls deploy` - doesn't build artifact for you, you have to run `mvn package` explicitly

Nevertheless, right now, if you try to run `sls deploy` without building, it will throw this:
```
sls-tests ➜ sls deploy
Serverless: Packaging service...

  Error --------------------------------------------------

  ENOENT: no such file or directory, open 'target/hello-dev.jar'

     For debugging logs, run again after setting the "SLS_DEBUG=*" environment variable.
```

While `sls invoke local -f hello` will throw this:
```
sls-tests ➜ ../serverless/bin/serverless invoke local -f hello
Building Java bridge, first invocation might take a bit longer.
java.lang.ClassNotFoundException: com.serverless.Handler

	at java.net.URLClassLoader.findClass(URLClassLoader.java:381)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:424)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:357)
	at java.lang.Class.forName0(Native Method)
	at java.lang.Class.forName(Class.java:348)
	at com.serverless.InvokeBridge.getInstance(InvokeBridge.java:53)
	at com.serverless.InvokeBridge.<init>(InvokeBridge.java:30)
	at com.serverless.InvokeBridge.main(InvokeBridge.java:86)
```

I think it would be better to unify error messages so `invoke local` would throw the same message as `deploy`, what do you think @pmuens ?

3. Sure

pmuens commented on lib/plugins/aws/invokeLocal/index.test.js:
```diff
@@ -352,12 +360,26 @@ describe('AwsInvokeLocal', () => {
+    });
+
+    it('throw error when using runtime other than Node.js or Python', () => {
+      awsInvokeLocal.options.functionObj.runtime = 'go';
```
Could we change this to `invalid-runtime` or something like that (so that it won't blow up if we add support for Go)?

pmuens commented on lib/plugins/aws/invokeLocal/index.test.js:
```diff
@@ -507,4 +529,189 @@ describe('AwsInvokeLocal', () => {
+
+    afterEach(() => {
+      serverless.cli.consoleLog.restore();
+      callJavaBridgeStub.restore();
```
Shouldn't this be `awsInvokeLocal.callJavaBridge.restore()`?

pmuens commented on lib/plugins/aws/invokeLocal/index.test.js:
```diff
@@ -507,4 +529,189 @@ describe('AwsInvokeLocal', () => {
+
+    it('should invoke callJavaBridge when bridge is built', () => {
+      const bridgePath = path.join(__dirname, 'java', 'target');
+      fs.mkdir(bridgePath);
```
This looks like a race condition (better use `fs.mkdirSync`).

pmuens commented on lib/plugins/aws/invokeLocal/index.test.js:
```diff
@@ -507,4 +529,189 @@ describe('AwsInvokeLocal', () => {
+            },
+          })
+        )).to.be.equal(true);
+        fs.rmdirSync(bridgePath);
```
Can we put this into an `afterEach()` to ensure that this is always removed even if the test fails?

pmuens commented on lib/plugins/aws/invokeLocal/index.test.js:
```diff
@@ -507,4 +529,189 @@ describe('AwsInvokeLocal', () => {
+      });
+
+      afterEach(() => {
+        callJavaBridgeMockedStub.restore();
```
Shouldn't this be `awsInvokeLocalMocked.callJavaBridge.restore()`?

RafalWilinski commented:
@pmuens Fixed!

RafalWilinski commented:
Thanks for guidance and fixes @pmuens ! Glad to see that merged finally 🙌

pmuens commented:
> Thanks for guidance and fixes @pmuens ! Glad to see that merged finally 🙌

Whoop, whoop! Thanks for taking the time to implement this @RafalWilinski!

Really great addition! 🙌 🎉
